scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala

```diff
@@ -937,9 +937,17 @@ class TypedPipeInst[T] private[scalding] (@transient inpipe: Pipe,
       // for us. So unwind until you hit the first filter, snapshot,
       // then apply the unwound functions
       case Some((tap, fields, Converter(conv))) =>
-        Execution.from(new Iterable[T] {
-          def iterator = m.openForRead(conf, tap).asScala.map(tup => conv(tup.selectEntry(fields)))
-        })
+        /*
+           * We create a Stream here so we are lazy with construction, but we don't read twice.
+           */
+        val tryStream = scala.util.Try(m.openForRead(conf, tap)
+          .asScala
+          .map(tup => conv(tup.selectEntry(fields)))
+          .toStream)
```

Comment:
Good point, when this using the KryoHadoop instantiator that I updated, so for Stream it should serialize by iterating all the data and forcing the stream. That might not be ideal, but since there is no contract that Iterable is lazy, this should be fine. If you send this Iterable to 1000 machines, the assumption should be that it is small.

If it is huge, then capturing in a closure is the wrong way to go (use a join or cross or manual closure that you create to force the laziness).